learning/prompts/documentation-lookup/31_advanced_batched_dofn.md

```diff
@@ -0,0 +1,35 @@
+Prompt:
+What is a Batched DoFn in Apache Beam?
+Response:
+In Apache Beam, a [Batched DoFn](https://beam.apache.org/documentation/programming-guide/#batched-dofns) is a [DoFn](https://beam.apache.org/documentation/programming-guide/#pardo) that can operate on batches of logical elements. These `DoFns` leverage popular Python vectorized libraries like [NumPy](https://numpy.org/), [SciPy](https://scipy.org/) and [pandas](https://pandas.pydata.org/) to efficiently process batches of elements in a `PCollection` rather than processing each element individually.
+
+Batched `DoFns` are only available in the Apache Beam Python SDK.
+
+You can provide element-wise, batched or both implementation for a `DoFn`
+
+```python
+class MultiplyByTen(beam.DoFn):
+  def process(self, element: np.int64) -> Iterator[np.int64]:
+    yield element * 10
+
+  def process_batch(self, batch: np.ndarray) -> Iterator[np.ndarray]:
+    yield batch * 10
+```
+Apache Beam runner will select the best implementation during pipeline execution.
+
+Then, when `MultiplyByTen` is applied to a `PCollection`, Beam recognizes that `np.ndarray` is an acceptable batch type to use in conjunction with `np.int64` elements.
+
+Apache Beam implicitly creates and explodes batches at the input and output boundaries. However, if Batched DoFns with equivalent types are chained together, this batch creation and explosion will be skipped making the compose batch transform more efficient.
```

Comment:
```suggestion
By default, Apache Beam implicitly buffers elements and creates batches on the input side, then explodes batches back into individual elements on the output side. However, if batched 'DoFn's with equivalent types are chained together, this batch creation and explosion process is skipped, and the batches are passed through for more efficient processing.

Here’s an example with chained ‘DoFn’s of equivalent types:

```

Reply:
This part in the documentation was provided AFTER an example of an implicit batch explosion (below). Suggest adding a similar example to the below first, before the "passed through" example, so that it is easier to understand what is 'chained' DoFn in contrast with 'non-chained' DoFn.

(p | beam.Create([1, 2, 3, 4]).with_output_types(np.int64)
   | beam.ParDo(MultiplyByTwo()) # Implicit buffering and batch creation
   | beam.Map(lambda x: x/3))  # Implicit batch explosion